### DISCUSSION TITLE
-extent'ing with Aspect Ratio Can Only Crop Images

### DISCUSSION THREAD

**Discussion by BlackXStar42** [ImageMagick/ImageMagick#4447] [Development]
As of version 7.1.0-2, there seems to be no way to extend an image to a certain aspect ratio in a single command. Both `-extent x:y` and `-extent x:y^` only _remove_ rows or columns of pixels to yield an image with the given aspect ratio. The caret modifier seems to be ignored for `x:y` altogether. While this totally makes sense for `-resize x:y` (as specifying a certain aspect ratio explicitly asks for distorting an imaging and there is only one way to do so), it would be very useful if we had an option to _add_ as many rows or columns of pixels necessary to yield an image with a certain aspect ratio. The caret would be a reasonable modifier to achieve this. One could also argue that the current behavior of `-extent x:y` is closer to the behavior of the greather-than modifier, so the new behavior should be achieved by using the less-than modifier. In this case, both modifiers should be allowed in the future for symmetry.

**Comment by GeeMack** [comment]
I agree it would be a useful addition to be able to increase the canvas to a given ratio. I've worked up a couple FX expressions to get that result when I need it, but it seems there could be a more direct method. When I was first aware of the use of the `X:Y` option, I thought it would be handy if `-crop X:Y` would reduce the canvas to the requested ratio, and if `-extent X:Y` would increase the canvas to contain the entire image inside the given ratio. I'd be interested in seeing if this suggestion gets any traction.

**Comment by snibgo** [comment]
I agree that this may be a useful addition, an easy way to add or subtract pixels to make a given aspect ratio.

I find "-extent" a bit confusing, as it may either add or subtract pixels [EDIT: or it can both add and subtract pixels]. For example:
```
magick -size 400x300 xc: -extent 100x100 info:

xc: XC 100x100 100x100+0+0 16-bit sRGB 0.000u 0:00.001
```
... so I'm not sure what to expect from "-extent 3:4".

**Reply by BlackXStar42** [reply]
As the command is called 'extent', it would be logical if the default behavior would be to _extend_ an image (if it is not already clear from the parameters whether to add or remove pixels) and that you need to add a modifier to modify the meaning of the command, e.g. to crop an image instead. But that's obviously not the way the devs thought when they made `x:y` available to the command. As the caret modifier for `-resize wxh` already does exactly the job to turn the fit-all-the-image-to-the-geometry behavior to the fill-all-the-geometry-with-the-image behavior, it would be intuitive to do the same for `-extent x:y`. On the other hand, the caret modifier is already ignored for `-resize x:y`, so... I'm not sure...

**Comment by snibgo** [comment]
> As the command is called 'extent', it would be logical if the default behavior would be to extend an image ...

I can see that argument, but that change would break existing scripts. Scripts may use "-extent 2:3", which removes pixels, so breaking that would be unwise.

"-resize 2:3" and "-extent 2:3" both remove pixels.

Modifiers caret and greater-than and less-than all make no change from the unmodified "-extent 2:3". These all remove columns or rows (never both). I doubt that any scripts currently use a modifier, so using one of these to add columns or rows is reasonable.

Reviewing https://imagemagick.org/script/command-line-processing.php#geometry , I think the caret modifier is the most logical for "extend (or resize) the image by one dimension to make the required aspect ratio".

So:
```
magick -size 100x100 -extent 2:3 info
```
... would create the result 67x100 pixels, as it currently does.
```
magick -size 100x100 -extent 2:3^ info
```
... would make the result 100x150.

(In Windows, caret is the escape character, so we need to escape it with another caret.)

**Comment by urban-warrior** [comment]
It is quite easy to support this request.  We can't use ^, however, as it is defined as `contain the given geometry given, (minimal dimensions given)`.  And internally it is defined as MinimumValue.  We need a new symbol to represent MaximumValue.  It could any symbol that is not already used in geometries: %, !, <, >, @, %  :, and ^.    What is your preference?

**Reply by BlackXStar42** [reply]
> It is quite easy to support this request. We can't use ^, however, as it is defined as `contain the given geometry given, (minimal dimensions given)`. ...

Strange, currently, IM just ignores the caret modifier in `-extent x:y^`. If its meaning is really as strict as you say, I would expect IM to report an error instead.

As for your question, I would probably go with the hash (#) if there isn't any important consideration for avoiding it. The character resembles 4 edges of an image that has been extended. Otherwise, I would probably go with an asterisk or a plus sign because they indicate _more_ like in _more pixels_.

**Comment by snibgo** [comment]
Hash seems good to me.

Plus isn't good because it has a different meaning in "2:3+0+0". Asterisk would be okay. But I prefer hash.

**Reply by BlackXStar42** [reply]
Right about the plus, I didn't think about it...

**Comment by fmw42** [comment]
How about the & or *?  Ampersand "the sign & standing for and" Has the connotation of more or larger.

**Reply by BlackXStar42** [reply]
I'm not convinced about ampersand, I always connect it with a bitwise conjunction, which tends to reduce the number of ones in the result.

**Comment by urban-warrior** [comment]
Hash it is.  Look for the enhancement in the Beta release within a week.

**Reply by GeeMack** [reply]
Excellent. We've had several users in the forums over the years ask how to make a proportional canvas to contain the entire image, frequently asking to make the result 1:1 square.

So if I understand the proposed modification, to make a square canvas that contains the entire image we'll be able to replace this...

    ...
    -extent "%[fx:max(w,h)]x%[fx:max(w,h)]" ^
    ...

With this...

    ...
    -extent "1:1#" ^
    ...

 And to make a 3:2 rectangular canvas that contains the image we'll replace this...

    ...
    -extent "%[w]x%[fx:w>h&&w/3>h/2?w/3*2:w<h&&h/3<w/2?w/2*3:h]" ^
    -extent "%[fx:w<h&&h/3>w/2?h/3*2:w>=h&&w/3<h/2?h/2*3:w]x%[h]" ^
    ...

With this...

    ...
    -extent "3:2#" ^
    ...

This will be an extremely useful modification. Thanks!

**Reply by BlackXStar42** [reply]
I didn't even know you can use calculations in the parameters...

**Reply by snibgo** [reply]
In IM v7, when using "magick", most options that use numbers can instead take %[fx:...] expressions. This includes options such as "-clone" and "-delete".

**Comment by urban-warrior** [comment]
Instead of hash, we have two existing operators we use for resizing: resize smaller for < and resize larger for >.  We could do something similar for crop and extent, e.g. `-extent "3:2>"` and `-extent "3:2<"`.  This provides for symmetrical behavior and should not break existing scripts.  Would this be the better option as compared to ^ and #?

**Reply by fmw42** [reply]
Yes, I like it. But would it break existing scripts that do not use any extra symbol for cropping via -extent or would that behavior continue?

**Comment by snibgo** [comment]
Ideally, it would also work for "-resize".

I have no objections to > and <.

**Comment by BlackXStar42** [comment]
Making the new behavior available to `-resize` is a good idea, of course.

Concerning the less-than and greater-than, as I pointed out, they were also one of my proposed set of modifiers. But then again, we have 2 considerations:
* Those 2 modifiers are conditional ones, while extending or resizing an image to a certain aspect ratio is unconditional. So strictly speaking, those modifiers would change their meaning when used with `x:y`. It is my understanding from the previous answers that this is undesirable.
* One of those modifiers will do the same as if you don't use any of them at all. This might look a bit asymmetrical, unless we remove the possibility to use `-extent x:y` or `-resize x:y` _without_ modifier, which would break existing scripts again. This is also undesirable.

So after thinking about it again, I'm not a big fan of less-than and greater-than anymore. In my opinion, we should still use a totally new modifier (hash), as, stricty speaking, it introduces a behavior which is not identical to any of the available modifiers. Summarizing, `-resize wxh` relates to `-resize wxh^` like `-extent x:y#` relates to `-extent x:y` as the respective first command will always contain the whole image, while the respective second command may remove parts.

**Comment by urban-warrior** [comment]
Clone the Gitlab main or wait 2 hours for the beta release and test the new '#' geometry modifier.  Post any feedback here.

**Comment by fmw42** [comment]
Seems to work fine for me for both # and ^ in the new beta

**Comment by BlackXStar42** [comment]
I really appreciate the new modifier, thanks!

**Comment by GeeMack** [comment]
Installed ImageMagick v7.1.0-14 Q16-HDRI x64, which includes this new addition, on Windows 10. I tested several variations of these "-extent" flags with portrait, landscape, and square input images. I tested with the gravity set to each position. I tried both the hash "#", to include the entire image, and the caret "^" or no flag to remove pixels. And I tested with floating point numbers like `-extent 2.75:1`.

Every combination I tried appears to work exactly as expected. Thanks for this very useful addition!
